### ISSUE TITLE
SA1402 (FileMayOnlyContainASingleClass) should prefer to keep the type matching the file name

### ISSUE THREAD

**Issue by sharwell** [DotNetAnalyzers/StyleCopAnalyzers#953]
Currently SA1402 reports a warning on all but the first top-level type declaration in a file. However, if a file contains a top-level type whose name matches the name of the file, that type should be the one without a warning reported, regardless of its position within the file.

### Example

**TypeName.cs**

```
namespace NamespaceName
{
    public interface ITypeName { }
    public class TypeName : ITypeName { }
}
```

#### Expected result

Warning SA1402 is reported on `ITypeName`, but not reported on `TypeName`.

#### Actual result

Warning SA1402 is reported on `TypeName`, but not reported on `ITypeName`.

### Impact

Since the code fix for SA1402 uses the diagnostics to determine which types need to move to new files, reporting a diagnostic on a type which should not move results in a violation of SA1649 (FileHeaderFileNameDocumentationMustMatchTypeName) appearing after the code fix is performed.

**Comment by dcooke28** [comment]
I'll look into this
